src/librustc_mir/hair/pattern/mod.rs

```diff
@@ -505,7 +505,16 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
                 let def = self.tables.qpath_def(qpath, pat.hir_id);
```

Comment:
I think it might be less fragile to check at the assignment of the tables field, whether `tables.tainted_by_errors` is true and then just not recurse.

Reply:
[Does this seem better](https://github.com/rust-lang/rust/pull/51696/commits/f5bacbc4852678b9d134e85d9a8f4011a45ed7f7)?

Reply:
Hmm no. You're right. your original thing seemed better. I thought we could check before calling `from_hir` or `new`, but I just checked and that seems messy.

r=me with your original thing